### ISSUE TITLE
Input for the Module creation command

### ISSUE THREAD

**Issue by janmedrek** [kyma-project/community#790]
Created on 2023-05-19 by Jan Medrek @janmedrek.

## Decision log

| Name | Description |
|-----------------------|------------------------------------------------------------------------------------|
| Title | Input for the Module creation command |
| Due date | 2023-05-31 |
| Status | Proposed on 2023-05-19, Accepted on 2023-05-31 |
| Decision type | Choice |
| Affected decisions | None |

## Context

Right now the Module creation is possible mainly via the CLI `alpha create module` command. It works properly with kubebuilder projects, but not as conveniently with the raw manifests or directories. Since we want to simplify the way of creating Modules we would like to shift towards a simple input strategy, which would include simple files that can be provided by any Module developer.

## Decision

The decision is to create the Module out of three YAML files:
- raw manifest, containing the K8S resources for the Module Operator
- default CR, containing the default Custom Resource that should be created when the Module is being enabled
- metadata file, which would include all the necessary metadata for the ModuleTemplate

As for the raw manifest and default CR, the structure is self-explanatory and enforced by the K8S and Module CRDs structure.
The metadata file that should be delivered along the manifests would have the following shape:
```
name: string              # required, the name of the Module
channel: string           # required, channel that should be used in the ModuleTemplate
version: string           # required, the version of the Module
resourceName: string      # optional, default={NAME}-{CHANNEL}, the name for the ModuleTemplate that will be created

internal: bool            # optional, default=false, determines whether the ModuleTemplate should have the internal flag or not
beta: bool                # optional, default=false, determines whether the ModuleTemplate should have the beta flag or not

labels:                   # optional, additional labels for the ModuleTemplate
  - string
annotations:              # optional, additional annotations for the ModuleTemplate
  - operator.kyma-project.io/doc-url={DOCUMENTATION_URL} # required, the link to the Module documentation
  - operator.kyma-project.io/support-url={SUPPORT_URL}   # required, the link where the User can get support regarding the Module
```

## Consequences

The Module creation/submission process will be simplified, developers will be able to easily provide input for the `alpha create module` command without depending on a specific tooling (such as kubebuilder).

**Comment by pbochynski** [comment]
1. Can you please provide an example file for any of the existing modules (e.g. keda)? 
2. As `doc-url` and `support-url` are stored as annotations maybe it would be better to put them directly in the annotation section. Easier implementation (conversion from camel case property to annotation), easier to understand for users (no mental mapping involved).

**Comment by janmedrek** [comment]
@pbochynski you are right, I've updated the structure. I am wondering, should we also include the Namespace in the file?

As for the example:
```
name: keda
channel: fast
version: v0.0.6-35a38a2
# resourceName is omitted, it will default to `keda-fast` (which is what it is named in the KCP already)

internal: false
beta: true

annotations:
  - operator.kyma-project.io/doc-url=https://github.com/kyma-project/keda-manager
  - operator.kyma-project.io/support-url=https://github.com/kyma-project/keda-manager/issues/new?assignees=&labels=&projects=&template=bug-report.md
```

**Comment by lindnerby** [comment]
@janmedrek Since we need a structured approach to non-kubebuilder project based module creation, this is a nice idea.
Probably we should make this the only alternative to it if the PROJ file can not be found, but we need some layout and name conventions then? Just thoughts:
* Raw resources can be one file `manifest.yaml` or a folder `manifest`?
* `default-cr.yaml` vs. parsing name or gvk from the metadata file and do lookup in files
* If we wan't to extend on the metadata file, what do you think of moving push registry url also there (instead of passing it via cmd flag) 
* Maybe format of this short metadata file should not be yaml, as a UX fluff, to have a clear distinction between resources and descriptive part. Maybe need to contain one `metadata.json`?
* Should the CRD also be clearly separated by some convention?
* Basically this will result in some kind of small custom project layout (don't know if this is easier than kubebuilder to be honest) but probably better to the alternative to reference everything from cmd flags like the resource part (manifest, cr, crd) and an additional metadata file. The command API will get even more convoluted and confusing.

**Comment by pbochynski** [comment]
> * Raw resources can be one file `manifest.yaml` or a folder `manifest`?

The manifest file or folder should be passed also referenced from the metadata file. This way Kyma CLI can have just one file as an input.

> * `default-cr.yaml` vs. parsing name or gvk from the metadata file and do lookup in files

I vote for file and reference the file from the metadata file. 

> * If we wan't to extend on the metadata file, what do you think of moving push registry url also there (instead of passing it via cmd flag)

There is a need to separate registry URL and credentials from the module content. The team that contributes module is not the one that has access to the registry where module template will be pushed.

> * Should the CRD also be clearly separated by some convention?

The lifecycle manager knows how to handle CRDs so it is not needed. If we allow to pass folder with manifests you are free to separate it or not.

> * Basically this will result in some kind of small custom project layout (don't know if this is easier than kubebuilder to be honest) but probably better to the alternative to reference everything from cmd flags like the resource part (manifest, cr, crd) and an additional metadata file. The command API will get even more convoluted and confusing.

Kubebuilder project gives only partial information required to build a module template and it is also not explicit. Please bear in mind that kubebuilder is not the only way to build an operator and having a small yaml/json file with explicit configuration will help to understand fully control module creation process.

**Comment by janmedrek** [comment]
Taking into the account above comments, the final version would be:
```
name: string              # required, the name of the Module
channel: string           # required, channel that should be used in the ModuleTemplate
version: string           # required, the version of the Module
resourceName: string      # optional, default={NAME}-{CHANNEL}, the name for the ModuleTemplate that will be created
manifest: string          # required, reference to the manifests, can be a name of a file or whole directory
security: string          # optional, name of the security scanners config file

internal: bool            # optional, default=false, determines whether the ModuleTemplate should have the internal flag or not
beta: bool                # optional, default=false, determines whether the ModuleTemplate should have the beta flag or not

labels:                   # optional, additional labels for the ModuleTemplate
  - string
annotations:              # optional, additional annotations for the ModuleTemplate
  - operator.kyma-project.io/doc-url={DOCUMENTATION_URL} # required, the link to the Module documentation
```

**Comment by pbochynski** [comment]
Decision accepted
